src/mds/StrayManager.cc

```diff
@@ -722,6 +723,7 @@ void StrayManager::migrate_stray(CDentry *dn, mds_rank_t to)
   ceph_tid_t tid = mds->issue_tid();
 
   auto req = make_message<MClientRequest>(CEPH_MDS_OP_RENAME);
+  req->set_reintegrate_op();
```

Comment:
Yeah, I didn't set any metadata on the inode. This PR just set a flag in the `rename client request`.

Reply:
If I understand the patch correctly, this only prevents repeated rename requests if the reintegration request is the one doing `path_traverse`. I think the robust way to resolve this is keeping track of the original request and waiting for a reply in `Server::handle_client_reply`.

Reply:
Actually the recursive `rename` was triggerred via `handle_client_rename() --> get_dentry_inode() --> link_remote()`. 

I haven't seen the case triggerred from `path_traverse()` yet. But from the code it's possible.

> I think the robust way to resolve this is keeping track of the original request and waiting for a reply in Server::handle_client_reply.

I didn't understand how could this work here.

The root cause is that when the first `reintegration` triggerred the MDS will send a `rename` client request to itself. And just while MDS handling or retrying the `rename` request each time will it trigger `handle_client_rename() --> get_dentry_inode() --> link_remote()`.

Reply:
> Actually the recursive `rename` was triggerred via `handle_client_rename() --> get_dentry_inode() --> link_remote()`.
> 
> I haven't seen the case triggerred from `path_traverse()` yet. But from the code it's possible.
> 
> > I think the robust way to resolve this is keeping track of the original request and waiting for a reply in Server::handle_client_reply.
> 
> I didn't understand how could this work here.
> 
> The root cause is that when the first `reintegration` triggerred the MDS will send a `rename` client request to itself. And just while MDS handling or retrying the `rename` request each time will it trigger `handle_client_rename() --> get_dentry_inode() --> link_remote()`.

When the MDS fires off a reintegration request (a rename RPC), it can save the reqid and a Context in some new map in `Server` and associate it (the reqid) with some new `reqid_t CDentry::reintegration_reqid` indicating the reintegration is in flight, don't issue a new one. When `Server::handle_client_reply` -- which only processes client requests from the MDS itself, currently just `rename` -- the server can complete the Context so the CDentry can clear `CDentry::reintegration_reqid`.

Makes sense?

Reply:
Hmm, yeah. This sound better. Let me have a try. Thanks @batrick

Reply:
Fixed it, but this will currently depend on https://github.com/ceph/ceph/pull/52199, please check the last 2 commits only.